### DISCUSSION TITLE
Dark Mode on BO

### DISCUSSION THREAD

**Discussion by mparvazi** [PrestaShop/PrestaShop#31016] [Ideas]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing features request](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3AImprovement%2CFeature%2CDeveloper+Feature+) and found no previous suggestion of this feature.

### Is your feature request related to a problem?

Dark mode has gained a lot of traction recently. Like Apple, for instance, has added dark mode to its iOS and MacOS operating systems. Windows and Google have done the same.

**Dark mode or no dark mode? That is the question.**
1. It’s cool and trendy (although that’s not a reason alone to do it)
2. It enhances accessibility by supporting users who are sensitive to eye strain in starkly bright themes.
3. It allows users to decide the most comfortable way to consume content while providing us a way to maintain control over the look and feel of things.
4. It helps to preserve battery life for devices with OLED screen where brighter colors consume more energy.
5. It’s extremely popular and appears to be going nowhere. It’s possible that your users who prefer a dark mode (like me!).

Sources:
[Material Design](https://material.io/design/color/dark-theme.html)
[CSS Tricks](https://css-tricks.com/a-complete-guide-to-dark-mode-on-the-web/)

### Describe the solution you'd like

Adding a button in header for toggling between light and dark mode.

![Screenshot 2021-12-30 at 15-58-33 Dashboard • dev](https://user-images.githubusercontent.com/85633460/147752636-bffb43d7-7126-4f4e-81e3-1811cb8de0c2.png)

![Screenshot 2022-01-01 at 18-05-43 Product • dark](https://user-images.githubusercontent.com/85633460/147857780-834c67f2-e6e8-4934-a3ff-c7601833aa6f.png)

![Screenshot 2022-01-01 at 18-36-17 Orders • dark](https://user-images.githubusercontent.com/85633460/147857778-9e5b7108-3272-43e9-99d3-baab6564659c.png)

**Dark Mode at the Operating System Level**

So far, we’ve used a button to toggle between light and dark mode but we can simply let the user’s operating system do that lifting for us. For example, many operating systems let users choose between light and dark themes directly in the system settings.

Fortunately, CSS has a [prefers-color-scheme](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme) media query which can be used to detect user’s system color scheme preferences. It can have three possible values: no preference, light and dark.

```
const prefersDarkScheme = window.matchMedia("(prefers-color-scheme: dark)");

if (prefersDarkScheme.matches) {
  document.body.classList.add("dark-theme");
} else {
  document.body.classList.remove("dark-theme");
}
```

### Do you plan to work on this feature?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [X] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by MatShir** [comment]
👀  @prestascott @NeOMakinG

**Comment by hibatallahAouadni** [comment]
Hello @mparvazi 

Thanks for bringing this up :pray: 
I think we've discussing about add a dark mode in BO with @NeOMakinG when @bulldesign sugguest to improve the BO design in this issue #23586

Can't wait for your PR :rocket: 

Thank you

**Comment by mparvazi** [comment]
## Current theming in BO
**_variables.scss**
```
$body-bg: #fff;
$body-color: #000;
```

**_content.scss**
```
body {
  color: $body-color;
  background-color: $body-bg;
}

```

## Dark/Light theming
### 1. Creating Sass map for theme colors

**_variables.scss**
```
$theme-light-colors:  (
    'body-bg': $body-bg,
    'body-color': $body-color,
);

$theme-dark-colors:  (
    'body-bg': invert($body-bg),
    'body-color': invert($body-color),
);
```

### 2. Converting colors map to CSS variables

**_light.scss**
```
:root.light {
    @each $color, $value in $theme-light-colors {
         --#{$color}: #{$value};
    }
}
```

**_dark.scss**
```
:root.dark {
    @each $color, $value in $theme-dark-colors {
         --#{$color}: #{$value};
    }
}
```

### 3. Using CSS variables instead of Sass variables in components

**_content.scss**
```
body {
    color: var(--body-color);
    background-color: var(--body-bg);
}
```
## Dark/Light toggling 
**theme.css**
```
:root.light {
    --body-bg: #fff;
    --body-color: #000;
}

:root.dark {
    --body-bg: #000;
    --body-color: #fff;
}
```
**header.tpl**
```
<html class="light">
    <body></body>
</html>
```

By toggling between `light` and `dark` class in `<html>`, corresponding CSS variables will be used.
Theme colors will be changed instantly without refreshing the page.

**Comment by NeOMakinG** [comment]
Funny I began something like this months ago and it looked almost like your screens

I would love to have such a feature in the BO honestly, wdyt @prestascott ?

He's not working atm but we really need his validation before

Also we would need to introduce a small dark/light toggler component (I think we need to use PHP because overwise we risk to have some color flash on loading)

**Comment by mparvazi** [comment]
Hello @NeOMakinG 

I tested in dashboard page by dark/light toggle button (without script and just with CSS vars).
There is not a big problem for color flash.
An interesting point is that after changing the color of the template, the color transitions works and the color of the template will be changed with transition.

I think this project should be started form [prestashop-ui-kit](https://github.com/PrestaShop/prestashop-ui-kit)
If upgraded to ‌Bootstrap 5, theming for [dark/light](https://prium.github.io/falcon/v3.6.0/modules/tables/advance-tables.html) is ready now on it (CSS vars used for theming)
If there is no plan for upgrade this [solution](https://github.com/PrestaShop/PrestaShop/issues/27149#issuecomment-1003454901) works fine in current theme.

Btw it would be a huge work, 2000 Sass variable have to be changed to CSS vars, and also mixin functions!

**Another problem is that IE doesn't support CSS vars, so there is two options:**
1. Duplicating rules (theme toggle doesn't work instantly):
```
color: white; //Fallback for IE
color: var(--white); //Used by other browsers
```
2. Using [script](https://github.com/nuxodin/ie11CustomProperties) for IE

**Comment by NeOMakinG** [comment]
> Hello @NeOMakinG
> 
> I tested in dashboard page by dark/light toggle button (without script and just with CSS vars). There is not a big problem for color flash. An interesting point is that after changing the color of the template, the color transitions works and the color of the template will be changed with transition.
> 
> I think this project should be started form [prestashop-ui-kit](https://github.com/PrestaShop/prestashop-ui-kit) If upgraded to ‌Bootstrap 5, theming for [dark/light](https://prium.github.io/falcon/v3.6.0/modules/tables/advance-tables.html) is ready now on it (CSS vars used for theming) If there is no plan for upgrade this [solution](https://github.com/PrestaShop/PrestaShop/issues/27149#issuecomment-1003454901) works fine in current theme.
> 
> Btw it would be a huge work, 2000 Sass variable have to be changed to CSS vars, and also mixin functions!
> 
> **Another problem is that IE doesn't support CSS vars, so there is two options:**
> 
> 1. Duplicating rules (theme toggle doesn't work instantly):
> 
> ```
> color: white; //Fallback for IE
> color: var(--white); //Used by other browsers
> ```
> 
> 2. Using [script](https://github.com/nuxodin/ie11CustomProperties) for IE

I think we don't care about IE, it will run out of support on June

We didn't plan to upgrade to Bootstrap 5 inside this major as there are big works on other subjects, but this could be in the next major

**Comment by prestascott** [comment]
Hello everyone,

Thank you @mparvazi for your request and I agree with all your points. Also, dark mode is more pleasant to use when you're using a product at night.

You did a great job here but I don't think that this topic is really prioritized. The Design Team must give you the correct colors in the dark theme after studying it because we need to give you colors that are accessible. This work will take some time but once done, you will be able to do the PR correctly. 

We can think of a toggle with Light mode and Dark mode. But also think about radio buttons for different options: Light mode, Dark mode, and Automatic (according to the Operating System) only if we want to add this option.
We can put the feature in Shop Parameters > General. I don't want to place this feature in order to make it obvious, it's still an appearance setting and somehow needs to be hidden a little bit.

For sure, I would also love to have this feature in the back office!

Wdyt?

**Comment by NeOMakinG** [comment]
@prestascott I think we shouldn't use the UX ✅  when you don't validate the subject but you responded, maybe we should just remove the `Waiting for UX`

**Comment by mparvazi** [comment]
Hello @prestascott 

I definitely agree with you!

Theming for dark mode is very different than light mode, so the principals should be prepared first.
The color depth in dark mode is very important:

<img width="200" src="https://user-images.githubusercontent.com/85633460/147915591-4309909d-250c-496e-a172-92715fe52143.png"></img>

As you can see in screenshots, I had some mistakes in coloring because in some parts layering didn't display correctly.

So I think the first step is creating a complete reference for coloring and color depth for layering.

Because I'm a dark theme lover, I can prepare a color scheme for start.
Then you can change it to your standard.
Please let me know if you're interested in to have it.

Btw dark theme principals for web applications is under study even by Google:
[Material System > Dark theme](https://material.io/design/color/dark-theme.html)

**Comment by mparvazi** [comment]
@NeOMakinG 

Wdyt about this:

Creating a plugin script for converting colors in `theme.css`
As like as [cssjanus](https://github.com/cssjanus/cssjanus) that works for RTL.

1- The script imports a color map:
```
dark (
    "#FFF": [ "#111" ],
    "#000": [ "#AAA", !box-shadow],
    "#000": [ "#FFF", color, border-color ],
)
```

2- Converts `theme.css` colors with regex and generate the `theme-dark.css`

For example automatic theming:
```
<link href="theme-dark.css" media="(prefers-color-scheme: dark)">
<link href="theme.css" media="(prefers-color-scheme: no-preference), (prefers-color-scheme: light)">
```

So there is no problem for color flash.

**Comment by NeOMakinG** [comment]
Mmmh, I would prefer to create a `theme-dark.scss` stylesheet with a dark version of the UIKit

It would be the cleaner solution without any third-party plugins
